### PR TITLE
Encrypt password on the frontend before sending the request. Closes #2

### LINKED ISSUE #2: Encrypt password on the frontend before sending the request
Security reasons

### PR DESCRIPTION
**Pull request by geakstr**
Encrypt user password with bcrypt and send this to server while login/register

Included new function `./js/utils/salt.js` for generate static salt by seed. Uses for generating salt by username

### CODE REVIEW

**Comment by mxstbr** [comment]
Thanks so much, I'll take a look soon!

**Comment by geakstr** [comment]
I think demo will stop working, because users localstorage will contain unencrypted passwords

**Comment by mxstbr** [comment]
Oh yeah actually, that might happen. What could we do to circumvent that?

**Comment by mxstbr** [comment]
I just tried it, and indeed as expected, previously registered password don't work. (e.g. the default username/password combination throws a "wrong password" error)

Any ideas how we could work around that?

**Comment by geakstr** [comment]
I don't see a good solution. Maybe set flag in localstorage: if flag not exist — clear storage. Or move from localstorage to in-memory storage.

**Comment by mxstbr** [comment]
Yeah, a flag might be a good way to go. Would you mind adding that?

**Comment by geakstr** [comment]
Ok, I'll do it :)

File: js/utils/fakeServer.js [hunk 1]
```diff
@@ -18,8 +19,11 @@ var server = {
     // populates the localStorage
     if (localStorage.users === undefined) {
       // Set default user
+      const AzureDiamond = "AzureDiamond";
+      const AzureDiamondSalt = genSalt(AzureDiamond);
+      const AzureDiamondPass = bcrypt.hashSync("hunter2", AzureDiamondSalt);
       users = {
-        "AzureDiamond": bcrypt.hashSync("hunter2", salt)
+        [AzureDiamond]: bcrypt.hashSync(AzureDiamondPass, salt)
```
Comment (mxstbr) [inline js/utils/fakeServer.js]: What do the square brackets here do? (`[AzureDiamond]`)

File: js/components/Form.react.js [hunk 2]
```diff
@@ -23,6 +23,7 @@ class LoginForm extends Component {
           <p className="form__error form__error--username-not-registered">This username does not exist.</p>
           <p className="form__error form__error--wrong-password">Wrong password.</p>
           <p className="form__error form__error--field-missing">Please fill out the entire form.</p>
+          <p className="form__error form__error--failed">Something wrong.</p>
```
Comment (mxstbr) [inline js/components/Form.react.js]: Lets change the error message to something a bit more friendly, maybe "Something went wrong, please try again!"?

File: js/utils/fakeServer.js (hunk 1 above)
Comment (geakstr) [inline js/utils/fakeServer.js]: This is a new feature in ES6 — dynamic properties. Can be here and not necessarily https://github.com/lukehoban/es6features#enhanced-object-literals

File: js/components/Form.react.js (hunk 2 above)
Comment (geakstr) [inline js/components/Form.react.js]: Ok

File: js/utils/salt.js [hunk 3]
```diff
@@ -0,0 +1,24 @@
+/**
+ * Generate 16 bytes salt for bcrypt by seed
+ *
+ * Should return the same salt for the same seed
+ *
+ * @param  {string}   seed The seed for salt
+ * @return {string}        The resulted salt
+ */
```
Comment (geakstr) [inline js/utils/salt.js]: I'm not sure that this function the best way to do this

File: js/utils/fakeServer.js (hunk 1 above)
Comment (mxstbr) [inline js/utils/fakeServer.js]: Ah didn't even know that, nice!

File: js/utils/salt.js (hunk 3 above)
Comment (mxstbr) [inline js/utils/salt.js]: What exactly does it do? I love the documentation anyway.

**Comment by mxstbr** [comment]
I just merged #6 and now we have some conflicts, would you mind resolving them? :+1:

File: js/utils/salt.js (hunk 3 above)
Comment (geakstr) [inline js/utils/salt.js]: As I understand salt on client can't be random and must be stable for username/password pair. This function generate stable compatible bcrypt salt. But I think my implementation is very naive and generated not "strong" salt, but I don't know how to do this better.

**Comment by geakstr** [comment]
@mxstbr fixed

File: js/utils/salt.js (hunk 3 above)
Comment (mxstbr) [inline js/utils/salt.js]: Why not `bcrypt.genSaltSync`?

Comment (geakstr) [inline js/utils/salt.js]: Because genSaltSync generates random salt. But client must send the same encrypted password each time, yes?

Comment (mxstbr) [inline js/utils/salt.js]: Yeah you're right – this is probably enough for client-side encryption then.

**Comment by mxstbr** [comment]
Thanks for the great PR, much appreciated!

**Comment by geakstr** [comment]
Very nice!

**Pull request merged**